presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -328,8 +328,8 @@ public QueryInfo createQuery(Session session, String query)
         QueryInfo queryInfo = queryExecution.getQueryInfo();
         queryMonitor.queryCreatedEvent(queryInfo);
 
-        queryExecution.addStateChangeListener(newValue -> {
-            if (newValue.isDone()) {
+        queryExecution.addQueryInfoStateChangeListener(newValue -> {
+            if (newValue.isFinalQueryInfo()) {
                 try {
                     QueryInfo info = queryExecution.getQueryInfo();
```

Comment:
the final query info is `newValue`, so this call is not needed.  I would rename `newValue` to `finalQueryInfo`